Internal pr

Kartik-Garg opened the pull request:
## Change Overview
Created CRD documentation using https://github.com/ahmetb/gen-crd-api-reference-docs/ , in response to the issue https://github.com/kanisterio/kanister/issues/1342 , added template folder, which contains files, which are used by the auto CRD generator tool, to create the CRD documentation, the documentation is named as "API.md" .

Signed-off-by: Kartik-Garg [kartikgarg938@gmail.com](mailto:kartikgarg938@gmail.com)

<!-- Insert PR description-->
Plan of action :

1. cloned git repo : https://github.com/ahmetb/gen-crd-api-reference-docs/
2. created an executable binary by running go build command
3. copied template folder from this repto to Kanister repo
4. ran the following command:
    `/path/to/gen-crd-api-reference-docs \ -config "/path/to/example-config.json" \ -api-dir "github.com/knative/build/pkg/apis/build/v1alpha1" \ -out-file docs.html`

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [x] :world_map: Documentation
- [ ] :robot: Test

## Issues <!-- to auto-close the issue, add the "fixes" keyword -->

- fixes #issue-number

## Test Plan

<!-- Will run prior to merging.-->
<!-- Include example how to run.-->

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

Kartik-Garg commented:
When the command `make docs` is ran:
`running DOCS_CMD in the containerized build environment Removing everything under '_build'... Scanning... Validating... The HTML pages are in _build/html. /usr/local/lib/python3.6/site-packages/sphinx_rtd_theme/search.html:20: RemovedInSphinx30Warning: To modify script_files in the theme is deprecated. Please insert a <script> tag directly in your theme instead. {{ super() }} make[1]: Entering directory '/home/kartik/Documents/OpenSource/Kanister/kanister' running CMD in the containerized build environment I1108 08:43:15.790940 1 main.go:132] parsing go packages in directory ./pkg/apis/cr/v1alpha1 I1108 08:43:33.314445 1 main.go:234] using package=github.com/kanisterio/kanister/pkg/apis/cr/v1alpha1 W1108 08:43:33.319322 1 main.go:445] not found external link source for type interface{} W1108 08:43:33.320886 1 main.go:445] not found external link source for type interface{} I1108 08:43:33.322448 1 main.go:170] written to API.md`

A snippet of API.md :
![Screenshot from 2022-11-08 14-52-53](https://user-images.githubusercontent.com/25133864/200795107-c410782d-fc97-4553-ba53-796ec8b38418.png)

PrasadG193 commented:
@Kartik-Garg let's not commit [gen-crd-api-reference-docs](https://github.com/kanisterio/kanister/pull/1727/files#diff-e650561fdf1a40c4a720bf94de4a5b3f699612be358cc5479a80907fe7d9ae8a) binary to the repo.
We can add this binary to the [build container](https://github.com/kanisterio/kanister/blob/master/Makefile#L65) that we use to build project. You can find the dockerfile here - https://github.com/kanisterio/kanister/blob/master/docker/build/Dockerfile

PrasadG193 commented:
We use different image for docs build - https://github.com/kanisterio/kanister/tree/master/docker/docs-build. Sorry for the confusion. We should add the binary in this build image

PrasadG193 commented:
First we need to decide when do we want to run this comment. And if this needs to be included in the public facing docs. Depending on that we can decide on the right docker image to add this binary.

Kartik-Garg commented:
> First we need to decide when do we want to run this comment. And if this needs to be included in the public facing docs. Depending on that we can decide on the right docker image to add this binary.

makes sense, so I have created a new target called crd_docs , so when `make crd_docs` is run the API.md is created then.

Kartik-Garg commented:
@PrasadG193  
I have added the additional changes:

1. Removed the executable binary from template folder, instead downloading the tar.gz asset and building binary in the docker file itself (https://github.com/Kartik-Garg/kanister/blob/1a55c417be803518c50c2df03cb96beb2dad1616/docker/build/Dockerfile#L26)
2. Added new target in the makefile which can be run explicitly whenever we want to create API.md file which contains documentation for API_CRD files : https://github.com/Kartik-Garg/kanister/blob/1a55c417be803518c50c2df03cb96beb2dad1616/Makefile#L211

Test Plan (manual testing) :

1. first I created a custom image locally which has similar code to build dockerfile : https://github.com/kanisterio/kanister/blob/master/docker/build/Dockerfile , in addition added code to create executable binary file which looks something like :
![Screenshot from 2022-11-11 15-12-01](https://user-images.githubusercontent.com/25133864/201315435-9b74dcd5-0bad-405e-9d85-0c0022ff9efc.png)

2. after pushing this image as: kartikgarg/crd:0.1 , created a new image variable in makefile : 
![Screenshot from 2022-11-11 15-11-17](https://user-images.githubusercontent.com/25133864/201315886-50a3c015-3afc-49a7-a498-cc557ee995d6.png)

3. Added the new target and command to run and test with the docker image:
![Screenshot from 2022-11-11 15-11-38](https://user-images.githubusercontent.com/25133864/201316274-56791298-f5ac-45fb-85b2-501b6d3e03a3.png)

4. Result is successful : 
![crd_docs_output](https://user-images.githubusercontent.com/25133864/201316489-be244a57-30dc-4a5b-940b-350f2e3db74d.png)

5. And API.md file is created: 
![Screenshot from 2022-11-11 15-32-57](https://user-images.githubusercontent.com/25133864/201316698-245c154e-4c58-493d-9332-0f92acee9b1c.png)

PrasadG193 commented:
Thanks a lot @Kartik-Garg for adding the Makefile target to build the docs.
We should move these docs to some directory. @pavannd1 @viveksinghggits I think the internal docs like API and Design docs should be in separate `/docs` folder and the website docs can be moved to `/website` dir.

PrasadG193 commented:
@pavannd1 does it make sense to add API docs section in docs.kanister.io and move this doc there? Something like - https://argocd-operator.readthedocs.io/en/latest/reference/api.html/

PrasadG193 commented on Makefile:
```diff
@@ -188,8 +188,7 @@ codegen:
                 make spelling && make html           \
-	   "
-
+	   "   
```
```suggestion
	   "

```

PrasadG193 commented on Makefile:
```diff
@@ -204,6 +203,25 @@ else
 	@/bin/bash -c $(DOCS_CMD)
 endif
 
+API_DOCS_CMD = "gen-crd-api-reference-docs		\
+				-config /usr/local/bin/example-config.json		\
+				-api-dir ./pkg/apis/cr/v1alpha1	\
+				-out-file API.md				\
+	   "
```
```suggestion
API_DOCS_CMD = "gen-crd-api-reference-docs		\
			-config /usr/local/bin/example-config.json		\
			-api-dir ./pkg/apis/cr/v1alpha1	\
			-out-file API.md				\
	   "

```

PrasadG193 commented on docker/build/Dockerfile:
```diff
@@ -22,6 +22,15 @@ COPY --from=golangci/golangci-lint:v1.50 /usr/bin/golangci-lint /usr/local/bin/
     && chmod +x /usr/local/bin/kind
 
+# adding crd documentation tool here, since docs-build does not have golang base to create binary from go file
+RUN wget -O gen-crd-api-reference-docs_linux_amd64.tar.gz https://github.com/ahmetb/gen-crd-api-reference-docs/archive/refs/tags/v0.3.0.tar.gz \
+    && tar xzfv gen-crd-api-reference-docs_linux_amd64.tar.gz  \
+    && mkdir crd_documentation \
+    && cp -r gen-crd-api-reference-docs-0.3.0/* ./crd_documentation \
+    && cd crd_documentation \
+    && go build \
+    && cp -r * /usr/local/bin/
```
Instead of buiding the binary from source, can't we download the prebuilt binary from the release page - https://github.com/ahmetb/gen-crd-api-reference-docs/releases/tag/v0.1.5?

Kartik-Garg replied:
The pre-built binary is outdated and there are certain fields, for which the older version is not able to create the documentation. Hence, had to go with the latest version. I had already commented regarding the same on the repo: 
![Screenshot from 2023-01-04 17-45-02](https://user-images.githubusercontent.com/25133864/210553178-e55ad93e-649d-474f-ac56-f5d67f4e53ba.png)
Since, there was no response from them, I chose to go with the version which covers everything for us.

Kartik-Garg replied:
I will look into the alternatives

Kartik-Garg replied:
@PrasadG193 I tried using some other alternatives, but did not get expected output, most of the output received were incomplete documentation. :/

Kartik-Garg commented on API.md:
```diff
@@ -0,0 +1,1363 @@
+<p>Packages:</p>
```
Alright I will do that

Kartik-Garg commented on Makefile:
```diff
@@ -204,6 +204,26 @@ else
 	@/bin/bash -c $(DOCS_CMD)
 endif
 
+API_DOCS_CMD = "gen-crd-api-reference-docs		\
+			-config /usr/local/bin/example-config.json		\
+			-api-dir ./pkg/apis/cr/v1alpha1	\
+			-out-file API.md				\
+	   "
```
Done

Kartik-Garg commented on template/pkg.tpl:
```diff
@@ -0,0 +1,49 @@
+{{ define "packages" }}
```
Since we are downloading the .tar.gz package, and it already has the template folder, I have completely removed the folder from our project, and I am using in-build flag to locate the template folder and utilize it.

PrasadG193 commented on docker/build/Dockerfile:
```diff
@@ -22,6 +22,15 @@ COPY --from=golangci/golangci-lint:v1.50 /usr/bin/golangci-lint /usr/local/bin/
     && chmod +x /usr/local/bin/kind
 
+# adding crd documentation tool here, since docs-build does not have golang base to create binary from go file
+RUN wget -O gen-crd-api-reference-docs_linux_amd64.tar.gz https://github.com/ahmetb/gen-crd-api-reference-docs/archive/refs/tags/v0.3.0.tar.gz \
+    && tar xzfv gen-crd-api-reference-docs_linux_amd64.tar.gz  \
+    && mkdir crd_documentation \
+    && cp -r gen-crd-api-reference-docs-0.3.0/* ./crd_documentation \
+    && cd crd_documentation \
+    && go build \
+    && cp -r * /usr/local/bin/
```
@Kartik-Garg we can achieve the same using this command. Could you please try?

```suggestion
RUN GOBIN=/usr/local/bin go install github.com/ahmetb/gen-crd-api-reference-docs@v0.3.0
```

Kartik-Garg replied:
@PrasadG193 it just gets the executable binary, but not the other required components like `config file` and `template folder` .

PrasadG193 replied:
Do we need other components? Looks like we are only copying the binary, right?

PrasadG193 replied:
If we just need a couple of files, lets maintain local copy instead of cloning whole repo just for this.

Kartik-Garg replied:
@PrasadG193 we do need just two other components, i.e. `config file` and the `template folder` .
Sure I will add a local copy of those and just download the binary using `RUN GOBIN` command.

PrasadG193 commented on docs/api_docs/template/members.tpl:
```diff
@@ -0,0 +1,48 @@
+{{ define "members" }}
```
Add comment about from where we are referring these files.
Also add license header of original repo if required.

Kartik-Garg replied:
@PrasadG193 , 

- Do I need to add comments on all of the files which we imported as to where they are being used?

- Where and how do I add the license?(I haven't done that before, so any insight would be helpful)

PrasadG193 replied:
@Kartik-Garg lets add to the files you have copied as-it-is. Just a comment should be enough

Kartik-Garg replied:
@PrasadG193 Added comments on the imported files, except the `config-json` file. Since, JSON files do not support comments.

PrasadG193 reviewed (approved):
LGTM

viveksinghggits commented on docs/api_docs/example-config.json:
```diff
@@ -0,0 +1,28 @@
+    "hideTypePatterns": [
+        "ParseError$",
+        "List$"
+    ],
```
@PrasadG193 should we change name of this file to just `config.json`?

Kartik-Garg replied:
@viveksinghggits I changed the file to just `config.json`

viveksinghggits reviewed (approved):
LGTM

viveksinghggits commented:
@Kartik-Garg
just to confirm this the documentation that is generated will be updated after every release right? Because we would be running make docs after every release.
Once new documentation is generated how would that we checked into the repo again, or thats not requried? Can you explain the workflow please?

Kartik-Garg commented:
> @Kartik-Garg just to confirm this the documentation that is generated will be updated after every release right? Because we would be running make docs after every release. Once new documentation is generated how would that we checked into the repo again, or thats not requried? Can you explain the workflow please?

@viveksinghggits The documentation would be generated every time we run the `make crd_docs` command. It would just create the documentation as `API.md` file. Once, the documentation is created and we need to update the existing one, we can just commit it and push it to repo, since the command creates documentation locally

viveksinghggits commented:
does that mean every time someone changes an API definition they will have to run this command and commit and push the documentation as well?

Kartik-Garg commented:
> does that mean every time someone changes an API definition they will have to run this command and commit and push the documentation as well?
Yes

viveksinghggits commented:
> > does that mean every time someone changes an API definition they will have to run this command and commit and push the documentation as well?
> 
> Yes

Let's add this workflow here so that people are aware https://github.com/kanisterio/kanister/blob/master/CONTRIBUTING.md#contributing-code

viveksinghggits commented:
Just another sub heading maybe, stating what to do when you update api types. I am assuming it's just the api types that we are concerned about here, right?

Kartik-Garg commented:
> > > does that mean every time someone changes an API definition they will have to run this command and commit and push the documentation as well?
> > 
> > 
> > Yes
> 
> Let's add this workflow here so that people are aware https://github.com/kanisterio/kanister/blob/master/CONTRIBUTING.md#contributing-code

@viveksinghggits Alright, can we add it under [this](https://github.com/kanisterio/kanister/blob/master/CONTRIBUTING.md#submitting-pull-requests) under the Contributing code? 
which would look something like this:
![Screenshot from 2023-02-22 22-24-44](https://user-images.githubusercontent.com/25133864/220700035-4d8a67f1-ba5b-4a34-8df9-40afacc1752c.png)

viveksinghggits commented:
I would rather prefer adding a subheading under to https://github.com/kanisterio/kanister/blob/master/CONTRIBUTING.md#contributing-code.


```
### Updating the API types

If your changes involve the Kanister API types as well, make sure you generate the API documentation using below command and then push the updated `API.md` as well.
```

Update wording and grammar please :smile: .

Kartik-Garg commented:
> Just another sub heading maybe, stating what to do when you update api types. I am assuming it's just the api types that we are concerned about here, right?

How does this look?
![Screenshot from 2023-02-22 22-33-37](https://user-images.githubusercontent.com/25133864/220702546-19defae5-a423-4483-923f-b434050842e2.png)

Kartik-Garg commented:
> I would rather prefer adding a subheading under to https://github.com/kanisterio/kanister/blob/master/CONTRIBUTING.md#contributing-code.
> 
> ```
> ### Updating the API types
> 
> If your changes involve the Kanister API types as well, make sure you generate the API documentation using below command and then push the updated `API.md` as well.
> ```
> 
> Update wording and grammar please smile .

Thanks a lot! @viveksinghggits 
I have made changes accordingly, please do let me know if everything seems alright

viveksinghggits commented:
Looks good. Thanks @Kartik-Garg :rocket:
